gc/api/include/OMR/IntrusiveList.hpp

```diff
@@ -0,0 +1,308 @@
+
+/*******************************************************************************
+ *  Copyright (c) 2018, 2018 IBM and others
+ *
+ *  This program and the accompanying materials are made available under
+ *  the terms of the Eclipse Public License 2.0 which accompanies this
+ *  distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ *  or the Apache License, Version 2.0 which accompanies this distribution and
+ *  is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ *  This Source Code may also be made available under the following
+ *  Secondary Licenses when the conditions for such availability set
+ *  forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ *  General Public License, version 2 with the GNU Classpath
+ *  Exception [1] and GNU General Public License, version 2 with the
+ *  OpenJDK Assembly Exception [2].
+ *
+ *  [1] https://www.gnu.org/software/classpath/license.html
+ *  [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ *  SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#if !defined(OMR_INTRUSIVELIST_HPP_)
+#define OMR_INTRUSIVELIST_HPP_
+
+#include <type_traits>
+#include <utility>
+#include <cassert>
+
+namespace OMR
+{
+
+template <typename T>
+class IntrusiveListNode;
+
+template <typename T>
+struct IntrusiveListTraits;
+
+template<typename T, typename Accessor>
+class IntrusiveList;
+
+template<typename T, typename Accessor>
+class IntrusiveListIterator;
+
+template<typename T, typename Accessor>
+class IntrusiveListConstIterator;
+
+/// Stores the next and previous pointers for elements in an IntrusiveList.
+/// Each element in an intrusive list must contain an IntrusiveListNode.
+/// The IntrusiveListTraits class is used to obtain the node from an element.
+template<typename T>
+class IntrusiveListNode
+{
+public:
+	constexpr IntrusiveListNode() noexcept : prev(nullptr), next(nullptr) {}
+
+	constexpr IntrusiveListNode(T* p, T* n) noexcept : prev(p), next(n) {}
+
+	/// assign previous, next to node.
+	void assign(T* p, T* n) noexcept
+	{
+		prev = p;
+		next = n;
+	}
+
+	// deactivate the node, and clear the next/prev pointers.
+	void clear() noexcept
```

Comment:
I do not believe noexcept is supported on the minimum version of Visual Studio OMR supports